File: smart-hub-direct-kt/src/main/kotlin/io/github/tassiLuca/hub/adapters/MockedHubManager.kt
package io.github.tassiLuca.hub.adapters

import io.github.tassiLuca.hub.application.LightingManager
import io.github.tassiLuca.hub.application.ThermostatManager
import io.github.tassiLuca.hub.core.LuminosityEntry
import io.github.tassiLuca.hub.core.TemperatureEntry
import kotlinx.coroutines.CoroutineScope
import kotlinx.coroutines.coroutineScope
import kotlinx.coroutines.flow.filter
import kotlinx.coroutines.flow.map
import kotlinx.coroutines.launch
import kotlin.coroutines.CoroutineContext

/** A mocked hub manager. */
class MockedHubManager(override val coroutineContext: CoroutineContext) : CoroutineScope {

    /** The dashboard. */
    val dashboard = SwingDashboardService()

    /** The thermostat manager. */
    val thermostatManager = ThermostatManager(dashboard, coroutineContext)

    /** The lighting system manager. */
    val lightingManager = LightingManager(dashboard, coroutineContext)

    /** The sensor source. */
    val sensorSource = GraphicalSource()

    /** Runs the mocked hub manager. */
    suspend fun run() = coroutineScope {
        val temperatureFlow = sensorSource.sensorEvents
            .filter { it is TemperatureEntry }
            .map { it as TemperatureEntry }
        val luminosityFlow = sensorSource.sensorEvents
            .filter { it is LuminosityEntry }
            .map { it as LuminosityEntry }
        launch { thermostatManager.run(temperatureFlow) }
        lightingManager.run(luminosityFlow)
    }
}

File: smart-hub-direct/src/main/scala/io/github/tassiLuca/hub/adapters/MockedHubManager.scala
package io.github.tassiLuca.hub.adapters

import gears.async.TaskSchedule.RepeatUntilFailure
import gears.async.{Async, AsyncOperations, ReadableChannel, Task}
import io.github.tassiLuca.hub.adapters.ui.DashboardUI
import io.github.tassiLuca.hub.core.{LuminosityEntry, TemperatureEntry}
import io.github.tassiLuca.rears.groupBy

import scala.language.postfixOps

/** A concrete hub manager, mocking sources with graphical views. */
class MockedHubManager(using Async, AsyncOperations):

  private val ui = DashboardUI()
  private val sensorsSource = GraphicalSource()
  private val thermostatManager = new MockedThermostatManager() with SwingDashboardService(ui)
  private val lightingManager = new MockedLightingManager() with SwingDashboardService(ui)

  def run(): Unit =
    thermostatManager.dashboard.updateSchedule(
      thermostatManager.thermostat.scheduler.schedule.map((d, t) => (s"${d._1}", s"${d._2}") -> s"$t"),
    )
    val channelBySensor = sensorsSource.publishingChannel.groupBy(_.getClass)
    Task {
      channelBySensor.read() match
        case Right((clazz, c)) if clazz == classOf[TemperatureEntry] =>
          thermostatManager.run(c.asInstanceOf[ReadableChannel[TemperatureEntry]])
        case Right((clazz, c)) if clazz == classOf[LuminosityEntry] =>
          lightingManager.run(c.asInstanceOf[ReadableChannel[LuminosityEntry]])
        case _ => ()
    }.schedule(RepeatUntilFailure()).run
    sensorsSource.asRunnable.run.await
